File: solution/f#/Nir-Kata-Tests/Parse.Dont.Validate/NIRProperties.fs
module NIRProperties

open FsCheck.Xunit
open NirGenerator
open Nir_Kata.Parse.Dont.Validate.NIRDomain.NIRDomain

let private shouldBeOk nir  =
    function
    | Ok parsedNir -> nir = parsedNir
    | _ -> failwith "Should be OK"

[<Property(Arbitrary = [| typeof<NirGenerator> |])>]
let ``roundtrip nir`` (nir: NIR) =
    nir.ToString()
    |> parseNIR
    |> shouldBeOk nir

File: solution/c#/Nir-Kata-Tests/Parse.Dont.Validate/NIRProperties.cs
using FsCheck;
using FsCheck.Xunit;
using Nir_Kata.Parse.Dont.Validate;
using static Nir_Kata.Parse.Dont.Validate.NIR;

namespace Nir_Kata_Tests.Parse.Dont.Validate
{
    public class NIRProperties
    {
        [Property(Arbitrary = new[] {typeof(NIRGenerator)})]
        public Property RoundTripNIR(NIR nir) =>
            ParseNIR(nir.ToString())
                .Contains(nir)
                .ToProperty();
    }
}